### PR TITLE
Logs invoked in new terminal

### LINKED ISSUE #273: Kubernetes Follow Logs command should invoke in new terminal
The shared terminal might be running another task. Additionally, since it's possible for the command to be run a long running task in the shared terminal, which could block future tasks. We should change the follow logs command to use a new terminal.

### PR DESCRIPTION
**Pull request by bnookala**
Fixes #273.

Logs will be invoked in a new terminal when invoked from the command palette with the 'Follow Logs' command.

Additionally, fixes a TODO I had noticed in which Terminal invocation would only work for Pods in the current namespace.

### CODE REVIEW

File: src/components/kubectl/logs.ts [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    Show,
+    Follow
+}
+
+const FOLLOW_LOGS_TERMINAL = "Kubernetes: Follow Logs";
+
+/**
+ * Fetches logs for a Pod. Handles use cases for fetching pods
+ * from an open document, or from the current namespace.
+ * @param kubectl
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: There's no point having JSDoc for parameters if it just states the parameter name, I feel?

File: src/components/kubectl/logs.ts [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        cmd = `${cmd} --namespace=${podSummary.namespace}`;
+    }
+
+    if (containerName) {
+        cmd = `${cmd} --container=${containerName}`;
+    }
+
+    if (displayMode === LogsDisplayMode.Follow) {
+        cmd = `${cmd} -f`;
+        kubectl.invokeInNewTerminal(cmd, FOLLOW_LOGS_TERMINAL);
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: Suggest follow terminals be labelled as to which pod/container they're following - otherwise you could have several of them and not know which is which.

File: src/components/kubectl/logs.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    kubectl.invokeInSharedTerminal(cmd);
+}
+
+/**
+ * Searches for a pod yaml spec from the open document
+ * or from the currently selected namespace.
+ * @param kubectl
+ * @param displayMode
+ */
+async function findPod(kubectl: Kubectl, displayMode: LogsDisplayMode) {
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: Consider declaring a return type.

File: src/components/kubectl/logs.ts [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        if (!isPod(obj)) {
+            return;
+        }
+
+        const podSummary = {
+            name: obj.metadata.name,
+            namespace: obj.metadata.namespace
+        };
+
+        getLogsForPod(kubectl, podSummary, displayMode);
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: Hmm, seems like we have a naming issue here - `findPodFromOpenDocument` implies it finds a pod, it says nothing about fetching logs.  Of course the naming issue is very likely inherited from `extension.ts` but it might be good to tidy this if we are doing a refactor.

File: src/components/kubectl/logs.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+        return;
+    } catch (ex) {
+        host.showErrorMessage("Could not retrieve Pod from Active Document");
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: "active document" (no capitals).

Actually, this sounds like we could retrieve a pod from the document, whereas I think it really means we couldn't retrieve _logs_ for the pod described by the active document... is that correct?  If so can we rephrase please?

File: src/components/kubectl/logs.ts [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+        [kuberesources.allKinds.pod],
+        { nameOptional: false },
+        (pod) => {
+
+            const podSummary: PodSummary = {
+                name: pod.split('/')[1],
+                namespace: namespace // should figure out how to handle namespaces.
+            };
+
+            getLogsForPod(kubectl, podSummary, displayMode);
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: Same comment about `findPod` function naming.  In the old code, `findPod` took a callback by the look of it; that callback now seems to have been hardwired into the function definitions.

File: src/extension.ts [hunk 7]
```diff
@@ -150,15 +151,15 @@ export async function activate(context): Promise<extensionapi.ExtensionAPI> {
         registerCommand('extension.vsKubernetesUseContext', useContextKubernetes),
         registerCommand('extension.vsKubernetesClusterInfo', clusterInfoKubernetes),
         registerCommand('extension.vsKubernetesDeleteContext', deleteContextKubernetes),
-        registerCommand('extension.vsKubernetesUseNamespace', (obj) => { useNamespaceKubernetes(kubectl, obj); } ),
+        registerCommand('extension.vsKubernetesUseNamespace', (explorerNode: explorer.KubernetesObject) => { useNamespaceKubernetes(kubectl, explorerNode); } ),
```
Comment (itowlson) [inline src/extension.ts]: I get this but I'd prefer it to be in a separate PR as it's not related to logs (and yeah I know I do this sometimes as well)... just makes it clearer when reading the diffs in six months' time.

File: src/components/kubectl/logs.ts (hunk 1 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: 👍

File: src/components/kubectl/logs.ts (hunk 2 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: 👍

File: src/components/kubectl/logs.ts (hunk 4 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: 👍

File: src/components/kubectl/logs.ts (hunk 5 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: 👍

File: src/components/kubectl/logs.ts (hunk 6 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: 👍

File: src/extension.ts (hunk 7 above)
Comment (bnookala) [reply src/extension.ts]: considering it's not a drastic functional change, and it's fairly small, it should be okay to leave it here.

File: src/components/kubectl/logs.ts [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+async function getLogsForExplorerPod(
+    kubectl: Kubectl,
+    explorerNode: ResourceNode,
+    displayMode: LogsDisplayMode
+) {
+    const namespace = explorerNode.namespace;
+    const podSummary = { name: explorerNode.id, namespace: namespace };
+    const container = await selectContainerForPod(podSummary);
+
+    // No container. How'd we get here. How does this Pod even work. What is going on.
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: _I_ think you should add "Inconceivable!!!" here but it is probably not worth stalling the PR just for a Princess Bride reference.

File: src/components/kubectl/logs.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+/**
+ * Finds a Pod from the open editor.
+ */
+async function logsForPodFromOpenDocument(kubectl: Kubectl, editor: vscode.TextEditor, displayMode: LogsDisplayMode) {
+    const text = editor.document.getText();
+    try {
+        const obj: {} = yaml.safeLoad(text);
+        if (!isPod(obj)) {
+            return;
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: I think you have changed the logic here.  Previously if the open document was not a pod, it would fall through to the prompt case (same as if there was no active document).  It looks like now if the open document is not a pod it will silently do nothing.  The old behaviour was intentional and we should keep it.

File: src/components/kubectl/logs.ts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+        const podSummary = {
+            name: obj.metadata.name,
+            namespace: obj.metadata.namespace
+        };
+
+        getLogsForPod(kubectl, podSummary, displayMode);
+        return;
+    } catch (ex) {
+        host.showErrorMessage("Could not retrieve logs for pod for active document.");
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: This seems wrong.  If I have, say, an application code file open, and ask for logs, I'd like to be able to select the pod running the application rather than have to close the document or open the pod YAML first.  Again, the old behaviour was to swallow the YAML parse exception and fall through to the 'select a container' case.

File: src/components/kubectl/logs.ts (hunk 9 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: fixed!

File: src/components/kubectl/logs.ts (hunk 8 above)
Comment (itowlson) [reply src/components/kubectl/logs.ts]: Actually I think this can happen if the pod contains multiple containers but the user cancels selection, so maybe we need to remove this comment - it might be misleading.

File: src/components/kubectl/logs.ts [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+ * Fetch logs from a Pod, when selected from the Explorer.
+ */
+async function getLogsForExplorerPod(
+    kubectl: Kubectl,
+    explorerNode: ResourceNode,
+    displayMode: LogsDisplayMode
+) {
+    const namespace = explorerNode.namespace;
+    const podSummary = { name: explorerNode.id, namespace: namespace };
+    const container = await selectContainerForPod(podSummary);
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: From here onwards this appears to be the same as `getLogsForPod` - could we just call that method?

File: src/components/kubectl/logs.ts [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+/**
+ * Finds a Pod from the open editor.
+ */
+async function logsForPodFromOpenDocument(kubectl: Kubectl, editor: vscode.TextEditor, displayMode: LogsDisplayMode) {
+    const text = editor.document.getText();
+    try {
+        const obj: {} = yaml.safeLoad(text);
+        if (!isPod(obj)) {
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: Consider flipping the condition - that way you don't need to have two places where you fall through to `logsForPodFromCurrentNamespace`.

File: src/components/kubectl/logs.ts [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+            // document doesn't desribe a pod.
+            return await logsForPodFromCurrentNamespace(kubectl, displayMode);
+        }
+
+        const podSummary = {
+            name: obj.metadata.name,
+            namespace: obj.metadata.namespace
+        };
+
+        getLogsForPod(kubectl, podSummary, displayMode);
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: `await`?

File: src/components/kubectl/logs.ts [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+        [kuberesources.allKinds.pod],
+        { nameOptional: false },
+        (pod) => {
+
+            const podSummary: PodSummary = {
+                name: pod.split('/')[1],
+                namespace: namespace // should figure out how to handle namespaces.
+            };
+
+            getLogsForPod(kubectl, podSummary, displayMode);
```
Comment (itowlson) [inline src/components/kubectl/logs.ts]: `await`?

Comment (bnookala) [reply src/components/kubectl/logs.ts]: fixed

File: src/components/kubectl/logs.ts (hunk 13 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: fixed

File: src/components/kubectl/logs.ts (hunk 12 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: fixed

File: src/components/kubectl/logs.ts (hunk 11 above)
Comment (bnookala) [reply src/components/kubectl/logs.ts]: fixed
